Macros/Macros.js

```diff
@@ -33,54 +35,163 @@ class FurnaceMacros {
+                    const tokenData = context.scene.getEmbeddedEntity("Token", remoteContext.tokenId);
+                    if (tokenData)
+                        context.token = new Token(tokenData, context.scene);
+                }
+            }
+            if (remoteContext.characterId)
+                context.character = game.actors.get(remoteContext.characterId) || null;
+        } else {
+            context.speaker = ChatMessage.getSpeaker();
+            context.actor = game.actors.get(context.speaker.actor);
+            context.token = canvas.tokens.get(context.speaker.token);
+            context.character = game.user.character;
         }
-      
-        const speaker = ChatMessage.getSpeaker();
-        const actor = game.actors.get(speaker.actor);
-        const token = canvas.tokens.get(speaker.token);
-        const character = game.user.character;
-        return { game: game, ui: ui, canvas: canvas, speaker, actor, token, character, event, sceneId: canvas.scene.id, args }
+        return context;
     }
-    
+
+    /**
+     * Defines whether a Macro can run as a GM.
+     * For security reasons, only macros authored by the GM, and not editable by users
+     * can be run as GM
+     */
+    canRunAsGM() {
+        const author = game.users.get(this.data.author);
+        const permissions = this.data.permission || {};
+        delete permissions[this.data.author];
+        return author && author.isGM && Object.values(permissions).every(p => p < CONST.ENTITY_PERMISSIONS.OWNER)
+    }
+
+    callScriptMacroFunction(context) {
+        const asyncFunction = this.data.command.includes("await") ? "async" : "";
+        return (new Function(`"use strict";
+            return (${asyncFunction} function ({speaker, actor, token, character, args, scene}={}) {
+                ${this.data.command}
+                });`))()(context);
+    }
+
     renderMacro(...args) {
         const context = FurnaceMacros.getTemplateContext(args);
-        if ( this.data.type === "chat" ) {
+        if (this.data.type === "chat") {
             if (this.data.command.includes("{{")) {
                 const compiled = Handlebars.compile(this.data.command);
-                return compiled(context, {allowProtoMethodsByDefault: true, allowProtoPropertiesByDefault: true})
+                return compiled(context, { allowProtoMethodsByDefault: true, allowProtoPropertiesByDefault: true })
             } else {
                 return this.data.command;
             }
         }
-        if ( this.data.type === "script" ) {
-            if ( !Macros.canUseScripts(game.user) )
+        if (this.data.type === "script") {
+            if (!Macros.canUseScripts(game.user))
                 return ui.notifications.warn(`You are not allowed to use JavaScript macros.`);
-            if (getProperty(this.data, "flags.furnace.gmexecute") && !game.user.isGM) {
-                return FurnaceMacros.requestRemoteExecute(this.name, ...args);
-            }
```

Comment:
Perhaps swap the order of these test to allow runasGM macros to be executed remotely even if the user does not have execute macro permission. (and remote the test in remote execute).
This would be good for trigger happy effects - like teleport and so on.

Reply:
It's a good idea, but I'm worried if a GM wants to use a 'runAsGM' macro but not allow anyone to use it, only those with script macro permissions. 
We can change it later, but for now, I don't want to give too much permissions and bypass a security feature as a feature then remove that feature later.